### PR TITLE
Autreg 52 update route53 dynamically

### PR DESCRIPTION
**Pull request by axthosarouris**

### CODE REVIEW

File: src/main/java/no/bibsys/lambda/api/handlers/UpdateStackRequestHandler.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+            createStacks(request,swaggerHubInfo);
+        }
+
+        if (request.getAction().equals(Action.DELETE)) {
+            deleteStacks(request,swaggerHubInfo);
+        }
+
+        System.out.println(request.toString());
+
+        ObjectMapper objectMapper = new ObjectMapper();
```
Comment (janbj) [inline src/main/java/no/bibsys/lambda/api/handlers/UpdateStackRequestHandler.java]: Better to use class variable?

File: run.py [hunk 2]
```diff
@@ -1,26 +1,31 @@
...
 
-import sys
 import subprocess
+import sys
 
 REPOSITORY = "repository"
 BRANCH = "branch"
 ACTION = "action"
 OWNER = "owner"
+SWAGGER_ORG = "swaggerOrg"
```
Comment (janbj) [inline run.py]: Starting to think we could need a short descriptive comment for each of these

File: src/main/java/no/bibsys/Application.java [hunk 3]
```diff
@@ -57,27 +62,32 @@ public static void main(String args[]) throws IOException {
         String action = System.getProperty("action");
         StringBuilder message = new StringBuilder(100);
         message.append("System property \"action\" is not set\n")
-            .append("Valid values: create,delete,update-role");
+            .append("Valid values: create,delete");
         Preconditions.checkNotNull(action, message.toString());
 
-        Application.run(repoOwner, repository, branch, action);
+        String swaggerOrg=System.getProperty("swaggerOrg");
```
Comment (janbj) [inline src/main/java/no/bibsys/Application.java]: Do you have a environment reader helper class?

File: src/main/java/no/bibsys/Application.java [hunk 4]
```diff
@@ -57,27 +62,32 @@ public static void main(String args[]) throws IOException {
...
     }
 
 
-
-    public void wipeStacks() {
+    public void wipeStacks(SwaggerHubInfo swaggerHubInfo, String networkZoneName)
+        throws IOException, URISyntaxException {
         checkNulls();
-        wiper.wipeStacks();
+        wiper.wipeStacks(swaggerHubInfo, networkZoneName);
```
Comment (janbj) [inline src/main/java/no/bibsys/Application.java]: Can we still wipe stacks if they don't use SwaggerHub and NetworkZones?

File: src/main/java/no/bibsys/apigateway/ServerInfo.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+            this.stage = stage;
+        }
+    }
+
+    /**
+     * It return  the Server URL in the form of  https://xxxxxxxx.execute-api.eu-west-1.amazonaws.com/{basePath}.
+     * It is intended for to be used by Swagger and SwaggerHub.
+     *
+     * @return the Server url to use in the OpenApi specification.
+     */
```
Comment (janbj) [inline src/main/java/no/bibsys/apigateway/ServerInfo.java]: Nice doc!

File: src/main/java/no/bibsys/lambda/api/handlers/UpdateStackRequestHandler.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+public class UpdateStackRequestHandler extends GithubHandler {
+
+
+    private transient SecretsReader secretsReader;
+    private final transient SwaggerHubInfo swaggerHubInfo;
+    private static final String API_KEY_HEADER="api-key";
+
+    public UpdateStackRequestHandler(Environment environment){
+        super(environment);
+        this.secretsReader = new SecretsReader("infrastructure", "buildbranch");
```
Comment (janbj) [inline src/main/java/no/bibsys/lambda/api/handlers/UpdateStackRequestHandler.java]: Use constants?

File: src/main/java/no/bibsys/lambda/api/handlers/UpdateStackRequestHandler.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+        if (request.getAction().equals(Action.CREATE)) {
+            createStacks(request,swaggerHubInfo);
+        }
+
+        if (request.getAction().equals(Action.DELETE)) {
+            deleteStacks(request,swaggerHubInfo);
+        }
+
+        System.out.println(request.toString());
```
Comment (janbj) [inline src/main/java/no/bibsys/lambda/api/handlers/UpdateStackRequestHandler.java]: Maybe use a logger here?

File: src/main/java/no/bibsys/lambda/deploy/handlers/DestroyHandler.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+        super();
+        this.environment=environment;
+
+    }
+
+    @Override
+    protected SimpleResponse processInput(DeployEvent input, String apiGatewayInputString,
+        Context context)
+        throws IOException, URISyntaxException {
+        Stage stage = Stage.currentStage();
```
Comment (janbj) [inline src/main/java/no/bibsys/lambda/deploy/handlers/DestroyHandler.java]: I'm thinking we should introduce some kind of task system and task interface when we want the init and destroyers to to more than one task

File: src/main/java/no/bibsys/lambda/deploy/handlers/Route53Updater.java [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+import no.bibsys.lambda.deploy.constants.NetworkConstants;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class Route53Updater {
+
+    public static final String ZONE_NAME_ENV = "ZONE_NAME";
+    public static final String REPOSITORY_NAME_ENV_VAR = "REPOSITORY";
+    public static final String BRANCH_NAME_ENV_VAR = "BRANCH";
+    public static final String CERTIFICATE_ARN = "REGIONAL_CERTIFICATE_ARN";
```
Comment (janbj) [inline src/main/java/no/bibsys/lambda/deploy/handlers/Route53Updater.java]: Maybe some more doc for these constants?

File: src/main/java/no/bibsys/lambda/deploy/handlers/SwaggerHubInfo.java [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+
+    private final transient String apiId;
+    private final transient String apiVersion;
+    private final transient String swaggerOrganization;
+
+
+    public SwaggerHubInfo(Environment environment) {
+        this.apiId = environment.readEnvOpt("API_ID").orElse("simple-api");
+        this.apiVersion = environment.readEnvOpt("API_VERSION").orElse("1.0");
+        this.swaggerOrganization = environment.readEnvOpt("SWAGGER_ORG").orElse("axthosarouris");
```
Comment (janbj) [inline src/main/java/no/bibsys/lambda/deploy/handlers/SwaggerHubInfo.java]: Hm, hardcoded values are not good. Maybe something we could provide in environment?

File: src/main/java/no/bibsys/lambda/deploy/handlers/SwaggerHubInfo.java [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+        return apiVersion;
+    }
+
+    public String getSwaggerOrganization() {
+        return swaggerOrganization;
+    }
+
+    public String getSwaggerAuth() throws IOException {
+
+        SecretsReader secretsReader = new SecretsReader("swaggerapikey", "swaggerapikey");
```
Comment (janbj) [inline src/main/java/no/bibsys/lambda/deploy/handlers/SwaggerHubInfo.java]: More magic strings that should be constants

File: src/main/java/no/bibsys/lambda/deploy/handlers/SwaggerHubUpdater.java [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+    }
+
+
+    public Optional<String> updateApiDocumentation()
+        throws IOException, URISyntaxException {
+        ApiDocumentationInfo apiDocInfo = constructApiDocumentationInfo(swaggerHubInfo.getApiVersion());
+        CloudFormationConfigurable config = new CloudFormationConfigurable(repository, branch);
+        Optional<String> jsonOpt = generateApiSpec(apiDocInfo, config);
+
+        System.out.println(jsonOpt.toString());
```
Comment (janbj) [inline src/main/java/no/bibsys/lambda/deploy/handlers/SwaggerHubUpdater.java]: Logger?

File: src/main/java/no/bibsys/lambda/deploy/handlers/SwaggerHubUpdater.java [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+import no.bibsys.apigateway.ApiGatewayApiInfo;
+import no.bibsys.cloudformation.CloudFormationConfigurable;
+import no.bibsys.cloudformation.Stage;
+import no.bibsys.swaggerhub.ApiDocumentationInfo;
+import no.bibsys.swaggerhub.SwaggerDriver;
+import org.apache.http.client.methods.HttpDelete;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.client.methods.HttpPost;
+
+public class SwaggerHubUpdater {
```
Comment (janbj) [inline src/main/java/no/bibsys/lambda/deploy/handlers/SwaggerHubUpdater.java]: Maybe a short doc about purpose and expected usage for this class?

File: src/main/java/no/bibsys/secrets/SecretsReader.java [hunk 14]
```diff
@@ -15,22 +15,29 @@
 
 
     private final transient AWSSecretsManager client;
+    private final transient String secretName;
+    private final transient String secretKey;
 
-    public SecretsReader(AWSSecretsManager client){
+    public SecretsReader(AWSSecretsManager client, String secretName, String secretKey) {
```
Comment (janbj) [inline src/main/java/no/bibsys/secrets/SecretsReader.java]: Small doc about purpose for this class?

File: src/main/java/no/bibsys/swaggerhub/SwaggerDriver.java [hunk 15]
```diff
@@ -41,13 +38,12 @@ public String executeGet(HttpGet get) throws IOException {
     }
 
 
-    public HttpGet getSpecificationVersionRequest(String apiVersion)
+    public HttpGet getSpecificationVersionRequest(String apiKey)
```
Comment (janbj) [inline src/main/java/no/bibsys/swaggerhub/SwaggerDriver.java]: Specificiation version for what?

File: src/main/java/no/bibsys/swaggerhub/SwaggerHubUrlFormatter.java [hunk 16]
```diff
@@ -39,14 +31,14 @@ public URI getRequestURL() {
...
 
 
-    private URI apiUri(String organization,String apiId,String version) throws URISyntaxException {
-        if (version != null && version.length() > 0) {
+    private URI apiUri(SwaggerHubInfo info,boolean includeApiVersion) throws URISyntaxException {
+        if (includeApiVersion) {
             return new URI(String
-                .format("https://api.swaggerhub.com/apis/%s/%s/%s", organization, apiId,
-                    version));
+                .format("https://api.swaggerhub.com/apis/%s/%s/%s", info.getSwaggerOrganization(), info.getApiId(),
```
Comment (janbj) [inline src/main/java/no/bibsys/swaggerhub/SwaggerHubUrlFormatter.java]: Hardcoded and repeated URL, maybe use constant or pattern in environment?

File: src/main/java/no/bibsys/utils/ResourceDestroyer.java [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+import com.amazonaws.services.apigateway.AmazonApiGateway;
+import com.amazonaws.services.apigateway.AmazonApiGatewayClientBuilder;
+import java.io.IOException;
+import java.net.URISyntaxException;
+import no.bibsys.cloudformation.Stage;
+import no.bibsys.lambda.deploy.handlers.Route53Updater;
+import no.bibsys.lambda.deploy.handlers.SwaggerHubInfo;
+import no.bibsys.lambda.deploy.handlers.SwaggerHubUpdater;
+
+public class ResourceDestroyer {
```
Comment (janbj) [inline src/main/java/no/bibsys/utils/ResourceDestroyer.java]: Add some more description

File: src/main/java/no/bibsys/utils/ResourceDestroyer.java [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+
+    public  void destroy()
+        throws IOException, URISyntaxException {
+
+        int response = swaggerHubUpdater.deleteApi();
+
+        this.route53Updater.deleteServerUrl();
+
+
+        System.out.println("Swagger response" + response);
```
Comment (janbj) [inline src/main/java/no/bibsys/utils/ResourceDestroyer.java]: Logger?

File: src/main/java/no/bibsys/utils/ResourceInitializer.java [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+import java.io.IOException;
+import java.net.URISyntaxException;
+import java.util.Optional;
+import no.bibsys.cloudformation.Stage;
+import no.bibsys.lambda.deploy.handlers.Route53Updater;
+import no.bibsys.lambda.deploy.handlers.SwaggerHubInfo;
+import no.bibsys.lambda.deploy.handlers.SwaggerHubUpdater;
+import no.bibsys.lambda.responses.SimpleResponse;
+
+public class ResourceInitializer {
```
Comment (janbj) [inline src/main/java/no/bibsys/utils/ResourceInitializer.java]: More doc telling what kind of resources this class handles and why

File: src/test/java/no/bibsys/cloudformation/ConfigurationTests.java [hunk 20]
```diff
@@ -18,7 +18,7 @@
     protected final String projectId;
     protected String branchName = "AUTREG-49_Delete_tables_from_DynamoDB_after_testing";
```
Comment (janbj) [inline src/test/java/no/bibsys/cloudformation/ConfigurationTests.java]: Should this refer to actual branches? What happens when they are removed?

File: src/test/java/no/bibsys/lambda/deploy/handlers/Route53UpdaterTest.java [hunk 21]
```diff
@@ -0,0 +1,119 @@
...
+        String branch = extractLocalBranch();
+        String certificateArn = "arn:aws:acm:eu-west-1:933878624978:certificate/b163e7df-2e12-4abf-ae91-7a8bbd19fb9a";
+
+        Route53Updater updater = new Route53Updater(zoneName, repository, branch, Stage.TEST,
+            AmazonApiGatewayClientBuilder.defaultClient());
+
+        Optional<ChangeResourceRecordSetsResult> result = updater.updateServerUrl(certificateArn);
+
+        assertThat(result, is(not(equalTo(Optional.empty()))));
+//        assertThat(result.get().getSdkHttpMetadata().getHttpStatusCode(), is(equalTo(200)));
```
Comment (janbj) [inline src/test/java/no/bibsys/lambda/deploy/handlers/Route53UpdaterTest.java]: Remove commented code

File: template.yml [hunk 22]
```diff
@@ -223,6 +225,10 @@ Resources:
           - codepipeline:PutJobFailureResult
           - codepipeline:PutJobSuccessResult
           - apigateway:*
+          - route53:ListHostedZones
+          - route53:ChangeResourceRecordSets
+          - apigateway:*
+
           Effect: Allow
           Resource: "*"
```
Comment (janbj) [inline template.yml]: Just an idea, but for the future do you think we could do a naming pattern to restrict this to resources belonging to our project, branch or stack?

File: src/main/java/no/bibsys/Application.java (hunk 3 above)
Comment (axthosarouris) [reply src/main/java/no/bibsys/Application.java]: I have, but these are system properties, i.e. java -DsystemProperty=systemValue

File: src/main/java/no/bibsys/apigateway/ApiGatewayApiInfo.java [hunk 23]
```diff
@@ -62,19 +60,41 @@ private String injectServerInfo(String openApiTemplate, ServerInfo serverInfo) {
 
     }
 
+
+    private String injectServerInfo(String openApiTemplate, ServerInfo serverInfo)
```
[the remark this answers is not in the document]
Comment (axthosarouris) [reply src/main/java/no/bibsys/apigateway/ApiGatewayApiInfo.java]: When we import the specification to API Gateway we miss some features, such as examples. This is specific to this project and we can create an issue for this problem in Github for future improvements

[the remark this answers is not in the document]
Comment (axthosarouris) [reply src/main/java/no/bibsys/apigateway/ApiGatewayApiInfo.java]: See below as well

File: src/main/java/no/bibsys/apigateway/ApiGatewayApiInfo.java [hunk 24]
```diff
@@ -62,19 +60,41 @@ private String injectServerInfo(String openApiTemplate, ServerInfo serverInfo) {
...
+            String removedVariables = yamlParser.writeValueAsString(root);
+            return removedVariables;
+
+
+        }
+
+    }
+
+
     private Optional<JsonNode> readOpenApiSpecFromAmazon(Map<String, String> requestParameters)
```
[the remark this answers is not in the document]
Comment (axthosarouris) [reply src/main/java/no/bibsys/apigateway/ApiGatewayApiInfo.java]: The only way to get the Server URL from an API is to read and parse the OpenAPI specification from API Gateway. That specification is for the time being incomplete, as it lacks examples.
I could not also succeed in sending responses, but it seems that you (@janbj ) have solved this problem. I suggest we create an issue in Github for this problem and deal with it later.

File: src/main/java/no/bibsys/apigateway/ApiGatewayApiInfo.java [hunk 25]
```diff
@@ -89,15 +109,23 @@ private String injectServerInfo(String openApiTemplate, ServerInfo serverInfo) {
...
+        Optional<String> apiStage = Optional.ofNullable(serversNode.get("variables"))
+            .map(var -> var.get("basePath"))
+            .map(basePath -> basePath.get("default"))
+            .map(deflt -> deflt.asText());
+        return apiStage;
     }
 
-    private Optional<RestApi> findRestApi(AmazonApiGateway apiGateway) {
-        List<RestApi> apiList = apiGateway
+    public Optional<RestApi> findRestApi() {
```
[the remark this answers is not in the document]
Comment (axthosarouris) [reply src/main/java/no/bibsys/apigateway/ApiGatewayApiInfo.java]: It was unclear. I have created a new class that lists resources for a specific Stack in order to find the REST API id and feed it into the ApiGatewayInfo class

File: src/main/java/no/bibsys/lambda/api/handlers/UpdateStackRequestHandler.java (hunk 7 above)
Comment (axthosarouris) [reply src/main/java/no/bibsys/lambda/api/handlers/UpdateStackRequestHandler.java]: I want to generally replace all system.outs with log after we merge the branch.

File: src/main/java/no/bibsys/lambda/deploy/handlers/SwaggerHubUpdater.java (hunk 12 above)
Comment (axthosarouris) [reply src/main/java/no/bibsys/lambda/deploy/handlers/SwaggerHubUpdater.java]: I want to look into logging as a separate thing

File: src/main/java/no/bibsys/lambda/deploy/handlers/DestroyHandler.java (hunk 8 above)
Comment (axthosarouris) [reply src/main/java/no/bibsys/lambda/deploy/handlers/DestroyHandler.java]: This depends on each project. The only requirement should be that we can call the Init lambda function

File: src/main/java/no/bibsys/secrets/SecretsReader.java (hunk 14 above)
Comment (axthosarouris) [reply src/main/java/no/bibsys/secrets/SecretsReader.java]: done

File: src/main/java/no/bibsys/lambda/deploy/handlers/SwaggerHubUpdater.java (hunk 13 above)
Comment (axthosarouris) [reply src/main/java/no/bibsys/lambda/deploy/handlers/SwaggerHubUpdater.java]: done

File: src/main/java/no/bibsys/utils/ResourceDestroyer.java (hunk 18 above)
Comment (axthosarouris) [reply src/main/java/no/bibsys/utils/ResourceDestroyer.java]: another issue for that

File: template.yml (hunk 22 above)
Comment (axthosarouris) [reply template.yml]: Good idea.

**Pull request merged**